docs(maintaining): use milestones over boards (#3228)
cdr/code-server: docs/MAINTAINING.md

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
# Maintaining

- [Maintaining](#maintaining)
  - [Workflow](#workflow)
    - [Milestones](#milestones)
    - [Project Boards](#project-boards)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# Maintaining

Current maintainers:

- @code-asher
- @oxy
- @jsjoeio

This document is meant to serve current and future maintainers of code-server, but also share openly our workflow for maintaining the project.

## Workflow

The workflow used by code-server maintainers is one that aims to be easy to understood by the community and easy enough for new maintainers to jump in and start contributing on day one.

### Milestones

We operate mainly using [Milestones](https://github.com/cdr/code-server/milestones). This was heavily inspired by our friends over at [vscode](https://github.com/microsoft/vscode).

> Here are the Milestones we use and how we use them:

Reviewer: To match other non-capitalized uses of milestones in the text:

```suggestion
Here are the milestones we use and how we use them:
```

> - "Backlog" -> Work not yet planned for a specific release.

Reviewer: So with this any un-assigned issues are essentially in triage state and once triaged we add to either "backlog" or "backlog candidates", right? Then eventually they move to "on deck" and then at some point to a milestone where they will finally be worked on.

Author: Exactly! I think you explained it perfectly. I'll add this too

- "On Deck" -> Work under consideration for upcoming milestones.

> - "Backlog Candidates" -> Work that is not yet accepted for the Backlog. We wait for the community to weigh in.

Reviewer: Just me nit picking :stuck_out_tongue: 

```suggestion
- "Backlog Candidates" -> Work that is not yet accepted for the "Backlog" milestone. We wait for the community to weigh in.
- ```

Reviewer: Or maybe just `for the backlog`. I think I'm being thrown off by the capitalization so it sounded like a proper noun to me.

> - "<Month> <Year>" -> Work to be done that month.

Reviewer: Is this where we landed? All good if so.

Author: Noooo - outdated! Will fix

### Project Boards

> We use project boards sparingly.

Reviewer: I think one way we could describe the use of project boards is something like "for projects or goals that span multiple milestones". So things like testing, clean up, etc fall under that description.

Author: Good call!

> Think of this as a place to put miscellaneous things (like testing, clean up stuff, etc). As a maintainer, random todos may come up here and there. This gives you a place to add notes without having to open a new issue (though that may be the better option — use your best judgement).

Reviewer: IMO if there's work to be done an issue always makes sense. Is the flow for creating a note significantly different than an issue?

Author: Main difference is notes aren't counted like Issues are. See creation flow here:
![2021-04-28 10 28 02](https://user-images.githubusercontent.com/3806031/116447205-9488cf80-a80c-11eb-8a3c-43d6015eb756.gif)

I just wanted an option to write a "todo" somewhere without significantly increasing our issue count (which some users use as a metric to indicate a project's healthiness). 

No strong opinion here. Happy to revise if you think everything should be an issue!

Author: > given that our release milestones function off of issues, then having it be a dedicated issue instead of a project note may make more sense

Discussed a bit more and going to change this wording to be "This gives you a place to add notes temporarily before opening a new issue"

## Workflow

The workflow used by code-server maintainers is one that aims to be easy to understood by the community and easy enough for new maintainers to jump in and start contributing on day one.

### Milestones

We operate mainly using [milestones](https://github.com/cdr/code-server/milestones). This was heavily inspired by our friends over at [vscode](https://github.com/microsoft/vscode).

Here are the milestones we use and how we use them:

- "Backlog" -> Work not yet planned for a specific release.
- "On Deck" -> Work under consideration for upcoming milestones.
- "Backlog Candidates" -> Work that is not yet accepted for the backlog. We wait for the community to weigh in.
- "<0.0.0>" -> Work to be done for that version.

With this flow, any un-assigned issues are essentially in triage state and once triaged are either "Backlog" or "Backlog Candidates". They will eventually move to "On Deck" (or be closed). Lastly, they will end up on a version milestone where they will be worked on.

### Project Boards

We use project boards for projects or goals that span multiple milestones.

> Think of this as a place to put miscellaneous things (like testing, clean up stuff, etc). As a maintainer, random todos may come up here and there. This gives you a place to add notes without having to open a new issue (though that may be the better option — use your best judgement).

Author: ```suggestion
Think of this as a place to put miscellaneous things (like testing, clean up stuff, etc). As a maintainer, random todos may come up here and there. This gives you a place to add notes temporarily before opening a new issue. Given that our release milestones function off of issues, we believe tasks should have dedicated issues.
```
